homeassistant/components/esphome/update.py

```diff
@@ -172,16 +165,18 @@ async def async_install(
     ) -> None:
         """Install an update."""
         async with self.hass.data.setdefault(KEY_UPDATE_LOCK, asyncio.Lock()):
-            device = self.coordinator.data.get(self._device_info.name)
+            coordinator = self.coordinator
+            api = coordinator.api
+            device = coordinator.data.get(self._device_info.name)
             assert device is not None
-            if not await self.coordinator.api.compile(device["configuration"]):
-                logging.getLogger(__name__).error(
-                    "Error compiling %s. Try again in ESPHome dashboard for error",
-                    device["configuration"],
-                )
-            if not await self.coordinator.api.upload(device["configuration"], "OTA"):
-                logging.getLogger(__name__).error(
-                    "Error OTA updating %s. Try again in ESPHome dashboard for error",
-                    device["configuration"],
-                )
-            await self.coordinator.async_request_refresh()
+            try:
+                if not await api.compile(device["configuration"]):
+                    raise HomeAssistantError(
+                        f"Error compiling {device['configuration']}; Try again in ESPHome dashboard for more information."
+                    )
+                if not await api.upload(device["configuration"], "OTA"):
+                    raise HomeAssistantError(
+                        f"Error updating via OTA: {device['configuration']}; Try again in ESPHome dashboard for more information."
+                    )
```

Comment:
The UI doesn't handle the failure well but thats not related to this PR (and this is the standard for reporting the error). At least you get feedback now.